File: common/test/java/org/openqa/selenium/ElementFindingTest.java
/*
Copyright 2007-2009 WebDriver committers
Copyright 2007-2009 Google Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package org.openqa.selenium;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;
import static org.openqa.selenium.Ignore.Driver.IPHONE;
import static org.openqa.selenium.Ignore.Driver.SELENESE;

import java.util.List;

public class ElementFindingTest extends AbstractDriverTestCase {

  public void testShouldReturnTitleOfPageIfSet() {
    driver.get(pages.xhtmlTestPage);
    assertThat(driver.getTitle(), equalTo(("XHTML Test Page")));

    driver.get(pages.simpleTestPage);
    assertThat(driver.getTitle(), equalTo("Hello WebDriver"));
  }

  public void testShouldNotBeAbleToLocateASingleElementThatDoesNotExist() {
    driver.get(pages.formPage);

    try {
      driver.findElement(By.id("nonExistantButton"));
      fail("Should not have succeeded");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  public void testShouldBeAbleToClickOnLinkIdentifiedByText() {
    driver.get(pages.xhtmlTestPage);
    driver.findElement(By.linkText("click me")).click();
    assertThat(driver.getTitle(), equalTo("We Arrive Here"));
  }

  public void testDriverShouldBeAbleToFindElementsAfterLoadingMoreThanOnePageAtATime() {
    driver.get(pages.formPage);
    driver.get(pages.xhtmlTestPage);
    driver.findElement(By.linkText("click me")).click();
    assertThat(driver.getTitle(), equalTo("We Arrive Here"));
  }

  public void testshouldBeAbleToClickOnLinkIdentifiedById() {
    driver.get(pages.xhtmlTestPage);
    driver.findElement(By.id("linkId")).click();
    assertThat(driver.getTitle(), equalTo("We Arrive Here"));
  }

  public void testShouldThrowAnExceptionWhenThereIsNoLinkToClickAndItIsFoundWithLinkText() {
    driver.get(pages.xhtmlTestPage);

    try {
      driver.findElement(By.linkText("Not here either"));
      fail("Should not have succeeded");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  public void testShouldfindAnElementBasedOnId() {
    driver.get(pages.formPage);

    WebElement element = driver.findElement(By.id("checky"));

    assertThat(element.isSelected(), is(false));
  }

  public void testShouldNotBeAbleTofindElementsBasedOnIdIfTheElementIsNotThere() {
    driver.get(pages.formPage);

    try {
      driver.findElement(By.id("notThere"));
      fail("Should not have succeeded");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @Ignore(SELENESE)
  public void testShouldBeAbleToFindChildrenOfANode() {
    driver.get(pages.xhtmlTestPage);
    List<WebElement> elements = driver.findElements(By.xpath("/html/head"));
    WebElement head = elements.get(0);
    List<WebElement> importedScripts = head.findElements(By.tagName("script"));
    assertThat(importedScripts.size(), equalTo(2));
  }

  @Ignore(SELENESE)
  public void testReturnAnEmptyListWhenThereAreNoChildrenOfANode() {
    driver.get(pages.xhtmlTestPage);
    WebElement table = driver.findElement(By.id("table"));
    List<WebElement> rows = table.findElements(By.tagName("tr"));

    assertThat(rows.size(), equalTo(0));
  }

  @Ignore(SELENESE)
  public void testShouldFindElementsByName() {
    driver.get(pages.formPage);

    WebElement element = driver.findElement(By.name("checky"));

    assertThat(element.getValue(), is("furrfu"));
  }

  public void testShouldFindElementsByClass() {
    driver.get(pages.xhtmlTestPage);

    WebElement element = driver.findElement(By.className("extraDiv"));
    assertTrue(element.getText().startsWith("Another div starts here."));
  }

  public void testShouldFindElementsByClassWhenItIsTheFirstNameAmongMany() {
    driver.get(pages.xhtmlTestPage);

    WebElement element = driver.findElement(By.className("nameA"));
    assertThat(element.getText(), equalTo("An H2 title"));
  }

  public void testShouldFindElementsByClassWhenItIsTheLastNameAmongMany() {
    driver.get(pages.xhtmlTestPage);

    WebElement element = driver.findElement(By.className("nameC"));
    assertThat(element.getText(), equalTo("An H2 title"));
  }

  public void testShouldFindElementsByClassWhenItIsInTheMiddleAmongMany() {
    driver.get(pages.xhtmlTestPage);

    WebElement element = driver.findElement(By.className("nameBnoise"));
    assertThat(element.getText(), equalTo("An H2 title"));
  }
  
  public void testShouldFindElementByClassWhenItsNameIsSurroundedByWhitespace() {
    driver.get(pages.xhtmlTestPage);
    
    WebElement element = driver.findElement(By.className("spaceAround"));
    assertThat(element.getText(), equalTo("Spaced out"));
  }
  
  @Ignore(SELENESE)
  public void testShouldFindElementsByClassWhenItsNameIsSurroundedByWhitespace() {
    driver.get(pages.xhtmlTestPage);
    
    List<WebElement> elements = driver.findElements(By.className("spaceAround"));
    assertThat(elements.size(), equalTo(1));
    assertThat(elements.get(0).getText(), equalTo("Spaced out"));
  }

  public void testShouldNotFindElementsByClassWhenTheNameQueriedIsShorterThanCandidateName() {
    driver.get(pages.xhtmlTestPage);

    try {
      driver.findElement(By.className("nameB"));
      fail("Should not have succeeded");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @Ignore(SELENESE)
  public void testShouldBeAbleToFindMultipleElementsByXPath() {
    driver.get(pages.xhtmlTestPage);

    List<WebElement> elements = driver.findElements(By.xpath("//div"));

    assertTrue(elements.size() > 1);
  }

  @Ignore(SELENESE)
  public void testShouldBeAbleToFindMultipleElementsByLinkText() {
    driver.get(pages.xhtmlTestPage);

    List<WebElement> elements = driver.findElements(By.linkText("click me"));

    assertTrue("Expected 2 links, got " + elements.size(), elements.size() == 2);
  }

  @Ignore(SELENESE)
  public void testShouldBeAbleToFindMultipleElementsByPartialLinkText() {
    driver.get(pages.xhtmlTestPage);

    List<WebElement> elements = driver.findElements(By.partialLinkText("ick me"));

    assertTrue(elements.size() == 2);
  }

  @Ignore(SELENESE)
  public void testShouldBeAbleToFindElementByPartialLinkText() {
    driver.get(pages.xhtmlTestPage);

    try {
      driver.findElement(By.partialLinkText("anon"));
    } catch (NoSuchElementException e) {
      fail("Expected element to be found");
    }
  }

  @Ignore(SELENESE)
  public void testShouldBeAbleToFindMultipleElementsByName() {
    driver.get(pages.nestedPage);

    List<WebElement> elements = driver.findElements(By.name("checky"));

    assertTrue(elements.size() > 1);
  }

  @Ignore(SELENESE)
  public void testShouldBeAbleToFindMultipleElementsById() {
    driver.get(pages.nestedPage);

    List<WebElement> elements = driver.findElements(By.id("2"));

    assertEquals(8, elements.size());
  }

  @Ignore(SELENESE)
  public void testShouldBeAbleToFindMultipleElementsByClassName() {
    driver.get(pages.xhtmlTestPage);

    List<WebElement> elements = driver.findElements(By.className("nameC"));

    assertTrue(elements.size() > 1);
  }

  // You don't want to ask why this is here
  public void testWhenFindingByNameShouldNotReturnById() {
    driver.get(pages.formPage);

    WebElement element = driver.findElement(By.name("id-name1"));
    assertThat(element.getValue(), is("name"));

    element = driver.findElement(By.id("id-name1"));
    assertThat(element.getValue(), is("id"));

    element = driver.findElement(By.name("id-name2"));
    assertThat(element.getValue(), is("name"));

    element = driver.findElement(By.id("id-name2"));
    assertThat(element.getValue(), is("id"));
  }

  @Ignore(SELENESE)
  public void testShouldFindGrandChildren() {
    driver.get(pages.formPage);
    WebElement form = driver.findElement(By.id("nested_form"));
    form.findElement(By.name("x"));
  }

  @Ignore(SELENESE)
  public void testShouldNotFindElementOutSideTree() {
    driver.get(pages.formPage);
    WebElement element = driver.findElement(By.name("login"));
    try {
      element.findElement(By.name("x"));
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @Ignore(SELENESE)
  public void testShouldReturnElementsThatDoNotSupportTheNameProperty() {
    driver.get(pages.nestedPage);

    driver.findElement(By.name("div1"));
    // If this works, we're all good
  }

  public void testShouldFindHiddenElementsByName() {
    driver.get(pages.formPage);

    try {
      driver.findElement(By.name("hidden"));
    } catch (NoSuchElementException e) {
      fail("Expected to be able to find hidden element");
    }
  }

  @Ignore(SELENESE)
  public void testShouldfindAnElementBasedOnTagName() {
    driver.get(pages.formPage);

    WebElement element = driver.findElement(By.tagName("input"));

    assertNotNull(element);
  }

  @Ignore(SELENESE)
  public void testShouldfindElementsBasedOnTagName() {
    driver.get(pages.formPage);

    List<WebElement> elements = driver.findElements(By.tagName("input"));

    assertNotNull(elements);
  }

  public void testFindingByCompoundClassNameIsAnError() {
    driver.get(pages.xhtmlTestPage);

    try {
      driver.findElement(By.className("a b"));
      fail("Compound class names aren't allowed");
    } catch (IllegalLocatorException e) {
      // This is expected
    }

    try {
      driver.findElements(By.className("a b"));
      fail("Compound class names aren't allowed");
    } catch (IllegalLocatorException e) {
      // This is expected
    }
  }

  @JavascriptEnabled
  public void testShouldBeAbleToClickOnLinksWithNoHrefAttribute() {
    driver.get(pages.javascriptPage);

    WebElement element = driver.findElement(By.linkText("No href"));
    element.click();

    // if any exception is thrown, we won't get this far. Sanity check
    assertEquals("Changed", driver.getTitle());
  }

  @Ignore({SELENESE})
  public void testShouldNotBeAbleToFindAnElementOnABlankPage() {
    driver.get("about:blank");

    try {
      // Search for anything. This used to cause an IllegalStateException in IE.
      driver.findElement(By.tagName("a"));
      fail("Should not have been able to find a link");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @Ignore({IPHONE})
  @NeedsFreshDriver
  public void testShouldNotBeAbleToLocateASingleElementOnABlankPage() {
    // Note we're on the default start page for the browser at this point.

    try {
      driver.findElement(By.id("nonExistantButton"));
      fail("Should not have succeeded");
    } catch (NoSuchElementException e) {
      // this is expected
    }
  }

  @JavascriptEnabled
  public void testRemovingAnElementDynamicallyFromTheDomShouldCauseAStaleRefException() {
    driver.get(pages.javascriptPage);

    RenderedWebElement toBeDeleted = (RenderedWebElement) driver.findElement(By.id("deleted"));
    assertTrue(toBeDeleted.isDisplayed());

    driver.findElement(By.id("delete")).click();

    try {
      toBeDeleted.isDisplayed();
      fail("Element should be stale at this point");
    } catch (StaleElementReferenceException e) {
      // this is expected
    }
  }

  public void testFindingALinkByXpathUsingContainsKeywordShouldWork() {
    driver.get(pages.nestedPage);

    try {
      driver.findElement(By.xpath("//a[contains(.,'hello world')]"));
    } catch (Exception e) {
      e.printStackTrace();
      fail("Should not have thrown an exception");
    }
  }

  @JavascriptEnabled
  public void testShouldBeAbleToFindAnElementByCssSelector() {
    driver.get(pages.xhtmlTestPage);
    if (!supportsSelectorApi()) {
      System.out.println("Skipping test: selector API not supported");
      return;
    }
    driver.findElement(By.cssSelector("div.content"));
  }

  @JavascriptEnabled
  public void testShouldBeAbleToFindAnElementsByCssSelector() {
    driver.get(pages.xhtmlTestPage);
    if (!supportsSelectorApi()) {
      System.out.println("Skipping test: selector API not supported");
      return;
    }
    driver.findElements(By.cssSelector("p"));
  }
}

File: javascript/node/test/element_finding_test.js
// Copyright 2013 Selenium committers
// Copyright 2013 Software Freedom Conservancy
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
//     You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

require('./lib/_bootstrap')(module);

var assert = require('assert'),
    By = require('selenium-webdriver').By,
    error = require('selenium-webdriver').error;

var test = require('./lib/testbase'),
    Browser = test.Browser,
    Pages = test.Pages;


test.suite(function(env) {
  var browsers = env.browsers,
      waitForTitleToBe = env.waitForTitleToBe;

  var driver;
  beforeEach(function() { driver = env.driver; });

  describe('finding elements', function() {

    test.it(
        'should work after loading multiple pages in a row',
        function() {
          driver.get(Pages.formPage);
          driver.get(Pages.xhtmlTestPage);
          driver.findElement(By.linkText('click me')).click();
          waitForTitleToBe('We Arrive Here');
        });

    describe('By.id()', function() {
      test.it('should work', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.id('linkId')).click();
        waitForTitleToBe('We Arrive Here');
      });

      test.it('should fail if ID not present on page', function() {
        driver.get(Pages.formPage);
        driver.findElement(By.id('nonExistantButton')).
            then(assert.fail, function(e) {
              assert.equal(error.ErrorCode.NO_SUCH_ELEMENT, e.code);
            });
      });

      test.ignore(browsers(Browser.ANDROID)).it(
          'should find multiple elements by ID even though that ' +
              'is malformed HTML',
          function() {
            driver.get(Pages.nestedPage);
            driver.findElements(By.id('2')).then(function(elements) {
              assert.equal(elements.length, 8);
            });
          });
    });

    describe('By.linkText()', function() {
      test.it('should be able to click on link identified by text', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.linkText('click me')).click();
        waitForTitleToBe('We Arrive Here');
      });

      test.it(
        'should be able to find elements by partial link text', function() {
          driver.get(Pages.xhtmlTestPage);
          driver.findElement(By.partialLinkText('ick me')).click();
          waitForTitleToBe('We Arrive Here');
        });

      test.it('should work when link text contains equals sign', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.linkText('Link=equalssign')).
            getAttribute('id').
            then(function(id) {
              assert.equal('linkWithEqualsSign', id);
            });
      });

      test.it('matches by partial text when containing equals sign',
        function() {
          driver.get(Pages.xhtmlTestPage);
          driver.findElement(By.partialLinkText('Link=')).
              getAttribute('id').
              then(function(id) {
                assert.equal('linkWithEqualsSign', id);
              });
        });

      test.it('works when searching for multiple and text contains =',
          function() {
            driver.get(Pages.xhtmlTestPage);
            driver.findElements(By.linkText('Link=equalssign')).
                then(function(elements) {
                  assert.equal(elements.length, 1);
                  return elements[0].getAttribute('id');
                }).
                then(function(id) {
                  assert.equal('linkWithEqualsSign', id);
                });
          });

      test.it(
          'works when searching for multiple with partial text containing =',
          function() {
            driver.get(Pages.xhtmlTestPage);
            driver.findElements(By.partialLinkText('Link=')).
                then(function(elements) {
                  assert.equal(elements.length, 1);
                  return elements[0].getAttribute('id');
                }).
                then(function(id) {
                  assert.equal('linkWithEqualsSign', id);
                });
          });

      test.it('should be able to find multiple exact matches',
          function() {
            driver.get(Pages.xhtmlTestPage);
            driver.findElements(By.linkText('click me')).
                then(function(elements) {
                  assert.equal(elements.length, 2);
                });
          });

      test.it('should be able to find multiple partial matches',
          function() {
            driver.get(Pages.xhtmlTestPage);
            driver.findElements(By.partialLinkText('ick me')).
                then(function(elements) {
                  assert.equal(elements.length, 2);
                });
          });

      test.ignore(browsers(Browser.OPERA)).
      it('works on XHTML pages', function() {
        driver.get(test.whereIs('actualXhtmlPage.xhtml'));
        driver.findElement(By.linkText('Foo')).getText().then(function(text) {
          assert.equal(text, 'Foo');
        });
      });
    });

    describe('By.name()', function() {
      test.it('should work', function() {
        driver.get(Pages.formPage);
        driver.findElement(By.name('checky')).
            getAttribute('value').
            then(function(value) {
              assert.equal('furrfu', value);
            });
      });

      test.it('should find multiple elements with same name', function() {
        driver.get(Pages.nestedPage);
        driver.findElements(By.name('checky')).then(function(elements) {
          assert.ok(elements.length > 1);
        });
      });

      test.it(
          'should be able to find elements that do not support name property',
          function() {
            driver.get(Pages.nestedPage);
            driver.findElement(By.name('div1'));
            // Pass if this does not return an error.
          });

      test.it('shoudl be able to find hidden elements by name', function() {
        driver.get(Pages.formPage);
        driver.findElement(By.name('hidden'));
        // Pass if this does not return an error.
      });
    });

    describe('By.className()', function() {
      test.it('should work', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.className('extraDiv')).
            getText().
            then(function(text) {
              var expected = 'Another div starts here.';
              var actual = text.substring(0, expected.length);
              assert.equal(actual, expected);
            });
      });

      test.it('should work when name is first name among many', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.className('nameA')).
            getText().
            then(function(text) {
              assert.equal(text, 'An H2 title');
            });
      });

      test.it('should work when name is last name among many', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.className('nameC')).
            getText().
            then(function(text) {
              assert.equal(text, 'An H2 title');
            });
      });

      test.it('should work when name is middle of many', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.className('nameBnoise')).
            getText().
            then(function(text) {
              assert.equal(text, 'An H2 title');
            });
      });

      test.it('should work when name surrounded by whitespace', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.className('spaceAround')).
            getText().
            then(function(text) {
              assert.equal(text, 'Spaced out');
            });
      });

      test.it('should fail if queried name only partially matches', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.className('nameB')).
            then(assert.fail, function(e) {
              assert.equal(error.ErrorCode.NO_SUCH_ELEMENT, e.code);
            });
      });

      test.it('should be able to find multiple matches', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElements(By.className('nameC')).then(function(elements) {
          assert.ok(elements.length > 1);
        });
      });

      test.it('does not permit compound class names', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.className('a b')).then(assert.fail, pass);
        driver.findElements(By.className('a b')).then(assert.fail, pass);
        function pass() {}
      });
    });

    describe('By.xpath()', function() {
      test.it('should work with multiple matches', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElements(By.xpath('//div')).then(function(elements) {
          assert.ok(elements.length > 1, elements.length);
        });
      });

      test.it('should work for selectors using contains keyword', function() {
        driver.get(Pages.nestedPage);
        driver.findElement(By.xpath('//a[contains(., "hello world")]'));
        // Pass if no error.
      });
    });

    describe('By.tagName()', function() {
      test.it('works', function() {
        driver.get(Pages.formPage);
        driver.findElement(By.tagName('input')).getTagName().
            then(function(name) {
              assert.equal(name, 'input');
            });
      });

      test.it('can find multiple elements', function() {
        driver.get(Pages.formPage);
        driver.findElements(By.tagName('input')).then(function(elements) {
          assert.ok(elements.length > 1);
        });
      });
    });

    describe('By.css()', function() {
      test.it('works', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElement(By.css('div.content'));
        // Pass if no error.
      });

      test.it('can find multiple elements', function() {
        driver.get(Pages.xhtmlTestPage);
        driver.findElements(By.css('p')).then(function(elements) {
          assert.ok(elements.length > 1);
        });
        // Pass if no error.
      });

      test.it('should be able to find element by compound CSS selector',
        function() {
          driver.get(Pages.xhtmlTestPage);
          assertAttributeIs(
              driver.findElement(By.css('div.extraDiv, div.content')),
              'class', 'content');
        });

      test.it('should be able to find multiple elements by compound selector',
          function() {
            driver.get(Pages.xhtmlTestPage);
            driver.findElements(By.css('div.extraDiv, div.content')).
                then(function(elements) {
                  assertAttributeIs(elements[0], 'class', 'content');
                  assertAttributeIs(elements[1], 'class', 'extraDiv');
                });
          });

      // IE only supports short version option[selected].
      test.ignore(browsers(Browser.IE)).
      it('should be able to find element by boolean attribute', function() {
        driver.get(test.whereIs(
            'locators_tests/boolean_attribute_selected.html'));
        assertAttributeIs(
            driver.findElement(By.css('option[selected="selected"]')),
            'value', 'two');
      });

      test.it(
          'should be able to find element with short ' +
              'boolean attribute selector',
          function() {
            driver.get(test.whereIs(
                'locators_tests/boolean_attribute_selected.html'));
            assertAttributeIs(
                driver.findElement(By.css('option[selected]')), 'value', 'two');
          });

      test.it(
          'should be able to find element with short boolean attribute ' +
              'selector on HTML4 page',
          function() {
            driver.get(test.whereIs(
                'locators_tests/boolean_attribute_selected_html4.html'));
            assertAttributeIs(driver.findElement(By.css('option[selected]')),
                'value', 'two');
          });

      function assertAttributeIs(element, attribute, expected) {
        element.getAttribute(attribute).then(function(attribute) {
          assert.equal(attribute, expected);
        });
      }
    });
  });
});
